rust/src/adt.rs

```diff
@@ -0,0 +1,909 @@
+// SPDX-License-Identifier: MIT
+use core::ffi::*;
+use core::mem::size_of;
+
+use crate::c_size_t;
+
+#[derive(Debug)]
+pub enum AdtError {
+    NotFound = -1,
+    BadOffset = -4,
+    BadPath = -5,
+    BadNCells = -14,
+    BadValue = -15,
+    BadLength = -20,
+}
+
+const ADT_ALIGN: usize = 4;
+const ADT_MAX_DEPTH: usize = 8;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct ADTNode {
+    property_count: u32,
+    child_count: u32,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct ADTProperty {
+    name: [c_char; 32],
+    size: u32,
+    value: [u8],
+}
+
+#[repr(C, packed(1))]
+pub struct ADTSegmentRanges {
+    phys: u64,
+    iova: u64,
+    remap: u64,
+    size: u32,
+    unk: u32,
+}
+
+/// Check that our address is properly aligned
+fn check_ptr(ptr: usize) -> Result<(), AdtError> {
+    if ptr % ADT_ALIGN != 0 {
+        Err(AdtError::BadOffset)
+    } else {
+        Ok(())
+    }
+}
+
+/// Determine if an ADT node's name is equal to some string bounded by a length.
+/// This is required to match the semantics expected when doing a full ADT path
+/// trace.
+fn node_names_equal(a: &str, b: &str, len: usize) -> bool {
```

Comment:
pre-slice b in the caller?

Reply:
Got rid of the need to slice the strings at all